R/makeNeighbourhoods.R

```diff
@@ -30,63 +32,104 @@
-            vertex.list.refined <- sapply(1:length(refined.vertices), FUN=function(X) neighbors(x, v=refined.vertices[X]))
-
-            return(refined.vertices)
-        } else if(isFALSE(refined)){
-            return(.sample_vertices(x, prop=prop, seed=seed, return.vertices=FALSE))
+        if(!is.matrix(reduced_dims) & isTRUE(refined)){
+            stop("No reduced dimensions matrix provided - required for refined sampling")
         }
-
+        graph <- x
+        X_reduced_dims <- reduced_dims
     } else{
         stop(paste0("Data format: ", class(x), " not recognised. Should be Milo or igraph"))
     }
+    random_vertices <- .sample_vertices(graph, prop, seed)
+    if (isFALSE(refined)) {
+        sampled_vertices <- random_vertices
+    } else if (isTRUE(refined)) {
+        vertex.knn <-
+            findKNN(
+                X = X_reduced_dims,
+                k = k,
+                subset = as.vector(random_vertices),
+                get.index = TRUE,
+                get.distance = FALSE
+            )
+        # knn_mat <-
+        #     Matrix(0, nrow = length(as.vector(random_vertices)), ncol = ncol(milo), sparse=TRUE)
+        # # knn_mat <- as(knn_mat, "sparseMatrix")
+        # for (ix in 1:nrow(knn_mat)) {
+        #     knn_mat[ix, vertex.knn$index[ix, ]] <- 1
+        # }
+        # ## Calculate avg profile of nearest neighbors
+        # nh_reduced_dims <- knn_mat %*% X_reduced_dims
+        # nh_reduced_dims <- t(apply(
+        #     nh_reduced_dims,
+        #     1,
+        #     FUN = function(x)
+        #         x / k
+        # ))
+       
+        nh_reduced_dims <- t(apply(vertex.knn$index, 1, function(x) colMedians(X_reduced_dims[x,])))
+        colnames(nh_reduced_dims) <- colnames(X_reduced_dims)
+        rownames(nh_reduced_dims) <- paste0('nh_', 1:nrow(nh_reduced_dims))
+        
+        ## Search nearest cell to average profile
+        # I have to do this trick because as far as I know there is no fast function to 
+        # search for NN between 2 distinct sets of points (here I'd like to search NNs of 
+        # nh_reduced_dims points among X_reduced_dims points). Suggestions are welcome
+        all_reduced_dims <- rbind(nh_reduced_dims, X_reduced_dims)
+        nn_mat <- BiocNeighbors::findKNN(all_reduced_dims,
+                                         k = nrow(nh_reduced_dims) + 1,
+                                         subset = rownames(nh_reduced_dims))[["index"]]
+        nn_mat_names <- apply(nn_mat, c(1,2), function(x) rownames(all_reduced_dims)[x])
+        i = 1
+        sampled_vertices <- rep("nh_0", nrow(nn_mat))
+        while (any(grepl(sampled_vertices, pattern = "nh_"))) {
+            update_ix <- grep(sampled_vertices, pattern="nh_")
+            sampled_vertices[update_ix] <- nn_mat_names[update_ix, i]
+            i <- i + 1
+        }
+        sampled_vertices <- match( sampled_vertices, rownames(X_reduced_dims))
+    } 
```

Comment:
Turn this into a separate function that can just be called after the conditional statement.